java/sage/io/RemoteSageFile.java

```diff
@@ -0,0 +1,555 @@
+  }
+
+  /**
+   * Open a file remotely via the media server.
+   *
+   * @param hostname The hostname of the media server hosting this file.
+   * @param name The full path and file to be opened.
+   * @param readonly Set the channel to be read only.
+   * @param transcodeMode Set the transcode mode.
+   * @throws IOException If there is an I/O related error
+   */
+  public RemoteSageFile(String hostname, String name, boolean readonly, String transcodeMode) throws IOException
+  {
+    this.hostname = hostname;
+    remoteFilename = name;
+    this.transcodeMode = transcodeMode;
+    this.readonly = readonly;
+    this.forceActive = false;
+    connect();
+  }
+
+  /**
+   * Open a file remotely via the media server.
+   *
+   * @param hostname The hostname of the media server hosting this file.
+   * @param file The file to be opened.
+   * @param readonly Set the channel to be read only.
+   * @param forceActive Always assume the file is active. Use this with care.
+   * @throws IOException If there is an I/O related error
+   */
+  public RemoteSageFile(String hostname, File file, boolean readonly, boolean forceActive) throws IOException
+  {
+    this(hostname, file.getPath(), readonly, forceActive);
+  }
+
+  /**
+   * Open a file remotely via the media server.
+   *
+   * @param hostname The hostname of the media server hosting this file.
+   * @param name The full path and file to be opened.
+   * @param readonly Set the channel to be read only.
+   * @param forceActive Always assume the file is active. Use this with care.
+   * @throws IOException If there is an I/O related error
+   */
+  public RemoteSageFile(String hostname, String name, boolean readonly, boolean forceActive) throws IOException
+  {
+    this.hostname = hostname;
+    remoteFilename = name;
+    this.transcodeMode = null;
+    this.readonly = readonly;
+    this.forceActive = forceActive;
+    connect();
+  }
+
+  private void connect() throws IOException
+  {
+    if (socket != null)
+      return;
+
+    Socket newSocket = new Socket();
+    newSocket.connect(new InetSocketAddress(hostname, 7818));
+    newSocket.setSoTimeout(30000);
+    newSocket.setTcpNoDelay(true);
+    socket = newSocket;
+
+    outStream = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream()));
+    inStream = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
+
+    if (transcodeMode != null && transcodeMode.length() > 0)
+    {
+      outStream.write(("XCODE_SETUP " + transcodeMode + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      String response = Sage.readLineBytes(inStream);
+      if (!"OK".equals(response))
+        throw new java.io.IOException("Error with remote transcode setup for " + transcodeMode + " of: " + response);
+    }
+
+    if (readonly)
+      outStream.write("OPENW ".getBytes(Sage.BYTE_CHARSET));
+    else
+      outStream.write("WRITEOPENW ".getBytes(Sage.BYTE_CHARSET));
```

Comment:
And if you do the uploadID thing, be sure to also release it when the file is closed (just request it again with zero size, that clears it on the server).